esphome/components/t6615/t6615.cpp

```diff
@@ -19,55 +19,66 @@ static const uint8_t T6615_COMMAND_ENABLE_ABC[] = {0xB7, 0x01};
 static const uint8_t T6615_COMMAND_DISABLE_ABC[] = {0xB7, 0x02};
 static const uint8_t T6615_COMMAND_SET_ELEVATION[] = {0x03, 0x0F};
 
+void T6615Component::send_ppm_command_() {
+  this->command_time_ = millis();
+  this->command_ = T6615Command::GET_PPM;
+  this->write_byte(T6615_MAGIC);
+  this->write_byte(T6615_ADDR_SENSOR);
+  this->write_byte(sizeof(T6615_COMMAND_GET_PPM));
+  this->write_array(T6615_COMMAND_GET_PPM, sizeof(T6615_COMMAND_GET_PPM));
+}
+
 void T6615Component::loop() {
-  if (!this->available())
+  if (this->available() < 5) {
+    if (this->command_ == T6615Command::GET_PPM && millis() - this->command_time_ > T6615_TIMEOUT) {
+      /* command got eaten, clear the buffer and fire another */
+      while (this->available())
+        this->read_byte(this->response_buffer_);
+      this->send_ppm_command_();
+    }
     return;
+  }
+
+  /* by the time we get here, we know we have at least five bytes in the buffer */
+  this->read_array(this->response_buffer_, 5);
 
   // Read header
-  uint8_t header[3];
-  this->read_array(header, 3);
-  if (header[0] != T6615_MAGIC || header[1] != T6615_ADDR_HOST) {
-    ESP_LOGW(TAG, "Reading data from T6615 failed!");
+  if (this->response_buffer_[0] != T6615_MAGIC || this->response_buffer_[1] != T6615_ADDR_HOST) {
+    ESP_LOGW(TAG, "Got bad data from T6615!");
```

Comment:
Historically we've sometimes seen fake copies of sensors that send different magic, it might be useful to log the mismatching part of the response:
```suggestion
    ESP_LOGW(TAG, "Got bad data from T6615, response started with %02X %02X!", response_buffer[0], response_buffer[1]);
```

Reply:
I separated out the magic and address to make it clearer what's different.